File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/MessageListAdapter.java
package io.euphoria.xkcd.app.impl.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import io.euphoria.xkcd.app.R;
import io.euphoria.xkcd.app.data.Message;

// TODO input bar
public class MessageListAdapter extends RecyclerView.Adapter<MessageListAdapter.ViewHolder>
                                implements DisplayListener {

    public interface InputBarListener {

        void onInputBarMoved(String oldParent, String newParent);

    }

    public enum InputBarDirection {
        UP, DOWN, LEFT, RIGHT, ROOT
    }

    // For logging
    private static final String TAG = "MessageListAdapter";

    // For view identification
    private static final int MESSAGE = 0;
    private static final int INPUT_BAR = 1;

    // The main data structure
    private final MessageForest data;
    // View of the input bar
    private final InputBarView inputBar;
    // MessageTree representation of the input bar
    private final MessageTree inputBarTree;

    private InputBarListener inputBarListener;

    public MessageListAdapter(MessageForest data, InputBarView inputBar) {
        this.data = data;
        this.inputBar = inputBar;
        if (data.has(MessageTree.CURSOR_ID)) {
            inputBarTree = data.get(MessageTree.CURSOR_ID);
        } else {
            inputBarTree = new MessageTree(null);
            data.add(inputBarTree);
        }
        data.setListener(this);
        inputBar.recycle();
        inputBar.setMessage(inputBarTree);
        setHasStableIds(true);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(
                Context.LAYOUT_INFLATER_SERVICE);
        switch (viewType) {
            case MESSAGE:
                MessageView mc = (MessageView) inflater.inflate(R.layout.template_message, parent, false);
                mc.setVisibility(View.INVISIBLE);
                return new ViewHolder(mc);
            case INPUT_BAR:
                inputBar.setVisibility(View.INVISIBLE);
                return new ViewHolder(inputBar);
            default:
                throw new IllegalArgumentException("Unknown view type: " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case MESSAGE:
                MessageView mc = (MessageView) holder.itemView;
                mc.recycle();
                final MessageTree mt = getItem(position);
                mc.setMessage(mt);
                mc.setTextClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        moveInputBarAround(mt);
                    }
                });
                mc.setCollapserClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        toggleCollapse(mt);
                    }
                });
                mc.setVisibility(View.VISIBLE);
                break;
            case INPUT_BAR:
                InputBarView ib = (InputBarView) holder.itemView;
                ib.setIndent(getItem(position).getIndent());
                ib.setVisibility(View.VISIBLE);
                break;
        }
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public MessageTree getItem(int position) {
        return data.get(position);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).getLongID();
    }

    @Override
    public int getItemViewType(int position) {
        MessageTree mt = getItem(position);
        return (mt.getMessage() == null) ? INPUT_BAR : MESSAGE;
    }

    public MessageForest getData() {
        return data;
    }

    public InputBarListener getInputBarListener() {
        return inputBarListener;
    }

    public void setInputBarListener(InputBarListener listener) {
        this.inputBarListener = listener;
    }

    private void dispatchInputBarMoved(String oldParentID, String newParentID) {
        if (inputBarListener != null) inputBarListener.onInputBarMoved(oldParentID, newParentID);
    }

    public int indexOf(MessageTree mt) {
        return data.findDisplayIndex(mt, true, false);
    }

    public MessageTree get(String id) {
        return data.get(id);
    }

    public MessageTree getTree(Message message) {
        return get((message == null) ? null : message.getID());
    }

    public MessageTree getParent(@NonNull MessageTree tree) {
        return data.getParent(tree);
    }

    public MessageTree getReply(MessageTree mt, int index) {
        return data.getReply(mt, index);
    }

    public MessageTree getSibling(MessageTree mt, int offset) {
        return data.getSibling(mt, offset);
    }

    public MessageTree add(@NonNull MessageTree mt) {
        return data.add(mt);
    }

    public MessageTree add(@NonNull UIMessage message) {
        return data.add(message);
    }

    public void remove(@NonNull MessageTree mt) {
        data.remove(mt, false);
    }

    // FIXME: Needs a better name.
    public void moveInputBarAround(@NonNull MessageTree mt) {
        String preferredID, alternateID;
        if (mt.getParent() == null) {
            preferredID = mt.getID();
            alternateID = null;
        } else {
            preferredID = mt.getParent();
            alternateID = mt.getID();
        }
        moveInputBar((preferredID.equals(inputBarTree.getParent())) ? alternateID : preferredID);
    }

    public void moveInputBar(String newParentID) {
        String oldParentID = inputBarTree.getParent();
        data.move(inputBarTree, get(newParentID), true);
        inputBar.setIndent(inputBarTree.getIndent());
        dispatchInputBarMoved(oldParentID, newParentID);
    }

    public boolean navigateInputBar(@NonNull InputBarDirection dir) {
        switch (dir) {
            case UP: // Predecessor of input bar or of closest parent to have one
                MessageTree node = inputBarTree;
                do {
                    MessageTree pred = getSibling(node, -1);
                    if (pred != null) {
                        moveInputBar(pred.getID());
                        return true;
                    }
                    node = getParent(node);
                } while (node != null);
                return false;
            case DOWN: // Most deeply nested first child of parent's successor, else parent
                if (inputBarTree.getParent() == null) return false;
                MessageTree par = getParent(inputBarTree);
                MessageTree succ = getSibling(par, 1);
                if (succ != null) {
                    // <n00b> cannot use infinite for loop idiom because of auto-formatter ;-; </noob>
                    while (true) {
                        MessageTree child = getReply(succ, 0);
                        if (child == null) break;
                        succ = child;
                    }
                    moveInputBar(succ.getID());
                } else {
                    moveInputBar(par.getParent());
                }
                return true;
            case LEFT: // The parent's parent
                if (inputBarTree.getParent() == null) return false;
                MessageTree parent = getParent(inputBarTree);
                moveInputBar(parent.getParent());
                return true;
            case RIGHT: // The immediate predecessor
                MessageTree pred = getSibling(inputBarTree, -1);
                if (pred == null) return false;
                moveInputBar(pred.getID());
                return true;
            case ROOT: // Just the root thread
                moveInputBar(null);
                return true;
        }
        return false;
    }

    public boolean tryEnsureVisible(@NonNull MessageTree mt, boolean expand) {
        return data.tryEnsureVisible(mt, expand);
    }

    public void toggleCollapse(@NonNull MessageTree mt, boolean newState) {
        data.setCollapsed(mt, newState);
    }

    public void toggleCollapse(@NonNull MessageTree mt) {
        data.toggleCollapsed(mt);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        public final BaseMessageView itemMessageView;

        public ViewHolder(BaseMessageView v) {
            super(v);
            itemMessageView = v;
        }

    }

}

File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/views/MessageListAdapter.kt
package io.euphoria.xkcd.app.impl.ui.views

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import io.euphoria.xkcd.app.R
import io.euphoria.xkcd.app.data.Message
import io.euphoria.xkcd.app.impl.ui.data.DisplayListener
import io.euphoria.xkcd.app.impl.ui.data.MessageForest
import io.euphoria.xkcd.app.impl.ui.data.MessageTree
import io.euphoria.xkcd.app.impl.ui.data.UIMessage

class MessageListAdapter(// The main data structure
    val data: MessageForest, // View of the input bar
    private val inputBar: InputBarView
) :
    RecyclerView.Adapter<MessageListAdapter.ViewHolder>(),
    DisplayListener {
    interface InputBarListener {
        fun onInputBarMoved(oldParent: String?, newParent: String?)
    }

    enum class InputBarDirection {
        UP, DOWN, LEFT, RIGHT, ROOT
    }

    // MessageTree representation of the input bar
    private var inputBarTree: MessageTree = data[MessageTree.Companion.CURSOR_ID] ?: MessageTree(null).also { data.add(it) }
    var inputBarListener: InputBarListener? = null
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val inflater = parent.context.getSystemService(
            Context.LAYOUT_INFLATER_SERVICE
        ) as LayoutInflater
        return when (viewType) {
            MESSAGE -> {
                val mc =
                    inflater.inflate(R.layout.template_message, parent, false) as MessageView
                mc.visibility = View.INVISIBLE
                ViewHolder(mc)
            }
            INPUT_BAR -> {
                inputBar.visibility = View.INVISIBLE
                ViewHolder(inputBar)
            }
            else -> throw IllegalArgumentException("Unknown view type: $viewType")
        }
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        when (holder.itemViewType) {
            MESSAGE -> {
                val mc = holder.itemView as MessageView
                mc.recycle()
                val mt = getItem(position)!!
                mc.setMessage(mt)
                mc.setTextClickListener(View.OnClickListener { moveInputBarAround(mt) })
                mc.setCollapserClickListener(View.OnClickListener { toggleCollapse(mt) })
                mc.visibility = View.VISIBLE
            }
            INPUT_BAR -> {
                val ib = holder.itemView as InputBarView
                ib.setIndent(getItem(position)!!.indent)
                ib.visibility = View.VISIBLE
            }
        }
    }

    override fun getItemCount() = data.size()

    fun getItem(position: Int): MessageTree? {
        return data[position]
    }

    override fun getItemId(position: Int): Long {
        return getItem(position)!!.longID
    }

    override fun getItemViewType(position: Int): Int {
        val mt = getItem(position)!!
        return if (mt.message == null) INPUT_BAR else MESSAGE
    }

    private fun dispatchInputBarMoved(
        oldParentID: String?,
        newParentID: String?
    ) {
        if (inputBarListener != null) inputBarListener!!.onInputBarMoved(oldParentID, newParentID)
    }

    fun indexOf(mt: MessageTree): Int {
        return data.indexOf(mt)
    }

    operator fun get(id: String): MessageTree? {
        return data[id]
    }

    fun getTree(message: Message): MessageTree? {
        return get(message.id)
    }

    fun getParent(tree: MessageTree): MessageTree? {
        return data.getParent(tree)
    }

    fun getReply(mt: MessageTree, index: Int): MessageTree? {
        return data.getReply(mt, index)
    }

    fun getSibling(mt: MessageTree, offset: Int): MessageTree? {
        return data.getSibling(mt, offset)
    }

    fun clear() {
        data.clear()
        data.add(inputBarTree)
    }

    fun add(mt: MessageTree): MessageTree? {
        return data.add(mt)
    }

    fun add(message: UIMessage): MessageTree? {
        return data.add(message)
    }

    fun remove(mt: MessageTree) {
        data.remove(mt, false)
    }

    // FIXME: Needs a better name.
    fun moveInputBarAround(mt: MessageTree?) {
        val preferredID: String?
        val alternateID: String?
        if (mt!!.parent == null) {
            preferredID = mt.id
            alternateID = null
        } else {
            preferredID = mt.parent
            alternateID = mt.id
        }
        moveInputBar(if (preferredID == inputBarTree.parent) alternateID else preferredID)
    }

    fun moveInputBar(newParentID: String?) {
        val oldParentID = inputBarTree.parent
        data.move(inputBarTree, newParentID?.let { get(it) }, ensureVisible = true)
        inputBar.setIndent(inputBarTree.indent)
        dispatchInputBarMoved(oldParentID, newParentID)
    }

    fun navigateInputBar(dir: InputBarDirection): Boolean {
        when (dir) {
            InputBarDirection.UP -> {
                var node: MessageTree? = inputBarTree
                do {
                    val pred = getSibling(node!!, -1)
                    if (pred != null) {
                        moveInputBar(pred.id)
                        return true
                    }
                    node = getParent(node)
                } while (node != null)
                return false
            }
            InputBarDirection.DOWN -> {
                if (inputBarTree.parent == null) return false
                val par = getParent(inputBarTree)!!
                var succ = getSibling(par, 1)
                if (succ != null) {
                    // <n00b> cannot use infinite for loop idiom because of auto-formatter ;-; </noob>
                    while (true) {
                        val child = getReply(succ!!, 0) ?: break
                        succ = child
                    }
                    moveInputBar(succ!!.id)
                } else {
                    moveInputBar(par.parent)
                }
                return true
            }
            InputBarDirection.LEFT -> {
                if (inputBarTree.parent == null) return false
                val parent = getParent(inputBarTree)
                moveInputBar(parent!!.parent)
                return true
            }
            InputBarDirection.RIGHT -> {
                val pred = getSibling(inputBarTree, -1) ?: return false
                moveInputBar(pred.id)
                return true
            }
            InputBarDirection.ROOT -> {
                moveInputBar(null)
                return true
            }
        }
        // when is exhaustive
    }

    fun tryEnsureVisible(mt: MessageTree, expand: Boolean): Boolean {
        return data.tryEnsureVisible(mt, expand)
    }

    fun toggleCollapse(mt: MessageTree, newState: Boolean) {
        data.setCollapsed(mt, newState)
    }

    fun toggleCollapse(mt: MessageTree) {
        data.toggleCollapsed(mt)
    }

    class ViewHolder(val itemMessageView: BaseMessageView) : RecyclerView.ViewHolder(
        itemMessageView
    )

    companion object {
        // For logging
        private const val TAG = "MessageListAdapter"

        // For view identification
        private const val MESSAGE = 0
        private const val INPUT_BAR = 1
    }

    init {
        data.setListener(this)
        inputBar.recycle()
        inputBar.setMessage(inputBarTree)
        setHasStableIds(true)
    }
}
